types/underscore/index.d.ts

```diff
@@ -813,54 +815,40 @@ declare module _ {
         unique: UnderscoreStatic['uniq'];
 
         /**
-        * Merges together the values of each of the arrays with the values at the corresponding position.
-        * Useful when you have separate data sources that are coordinated through matching array indexes.
-        * If you're working with a matrix of nested arrays, zip.apply can transpose the matrix in a similar fashion.
-        * @param arrays The arrays to merge/zip.
-        * @return Zipped version of `arrays`.
-        **/
-        zip(...arrays: any[][]): any[][];
-
-        /**
-        * @see _.zip
-        **/
-        zip(...arrays: any[]): any[];
-
-        /**
-        * The opposite of zip. Given a number of arrays, returns a series of new arrays, the first
-        * of which contains all of the first elements in the input arrays, the second of which
-        * contains all of the second elements, and so on. Use with apply to pass in an array
-        * of arrays
-        * @param arrays The arrays to unzip.
-        * @return Unzipped version of `arrays`.
-        **/
-        unzip(...arrays: any[][]): any[][];
-
-        /**
-        * Converts arrays into objects. Pass either a single list of [key, value] pairs, or a
-        * list of keys, and a list of values.
-        * @param keys Key array.
-        * @param values Value array.
-        * @return An object containing the `keys` as properties and `values` as the property values.
-        **/
-        object<TResult extends {}>(
-            keys: _.List<string>,
-            values: _.List<any>): TResult;
+         * Merges together the values of each of the `lists` with the values at
+         * the corresponding position. Useful when you have separate data
+         * sources that are coordinated through matching list indexes. If
+         * you're working with a matrix of nested lists, this can be used to
+         * transpose the matrix.
+         * @param lists The lists to zip.
+         * @returns The zipped version of `lists`.
+         **/
+        zip(...lists: List<any>[]): any[][];
```

Comment:
Once TS3.1 becomes the minimum version this September, it will be possible to update this to the below to provide a more useful result based on the value types in the provided lists.

```TS
zip<T extends _.List<any>[]>(...lists: T): { [K in keyof T]: T[K] extends _.List<infer TItem> ? TItem : never }[];`
```

Reply:
Given that it will be just two months, I wouldn't object to making the upgrade now already, either.

For my understanding: does the mapped type part of the return type result in a tuple where each member may potentially have a different type? That would be really nice, but I suspect it would require the tuple spread syntax in TS 4.

Reply:
Given a `List<string>`, a `List<number>`, and a `List<boolean>`, the result will be a `[string, number, boolean][]`. This takes advantage of [this feature](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-1.html#mapped-types-on-tuples-and-arrays) introduced in TS3.1, which allows one to enumerate over the types in a tuple (in this case the tuple would be `[List<string>, List<number>, List<boolean>]`).

The issue with pushing up past the minimum version early is that I'd have to also push up all of this package's dependencies early (including Backbone and everything that depends on Backbone) so I don't think it's feasible to do this at this time.

Reply:
To provide more detail here, when pushing the minimum version up past TS3.0, `npm run test` currently provides the following error: `Error: backbone depends on underscore but has a lower required TypeScript version.`.

Looking again at the [DT TS version support window](https://github.com/DefinitelyTyped/DefinitelyTyped#older-versions-of-typescript-29-and-earlier), though, I think I might have been reading it wrong; it seems like possibly TS3.0 will go out of support some time this month (at the end of it maybe?), which might possibly make TS3.1 the new minimum supported version some time in the next week or so.

It might be worth sticking this one on the back burner for a few days and coming back to it to see whether it's possible to use the new stuff sooner rather than later; let me know what your thoughts are here. I apologize for possibly opening this one up a little ahead of its optimal opening time.

Reply:
Really nice that mapped tuples are a thing as of 3.1.

I don't have a strong opinion on whether to upgrade now or later, delay the PR or not, etcetera.

Reply:
I had changed plans a bit to put `invoke` all by itself in a PR at the very end since I thought it was the only one that would need 3.0+ stuff before September, but I think what I'll do here instead is add more basic invoke changes to #46352 to wrap up Collections, make all the remaining changes for this PR set with 2.8 as the minimum version, and then do a PR after cleanup where I bump the version up to whatever works and make changes as appropriate.